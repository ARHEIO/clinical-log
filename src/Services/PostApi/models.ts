export interface Post {
  name: Name; // name of user posting
  time: string; // date of posting
  content: string;
}

export interface Name {
  firstName: string;
  lastName: string;
}

export interface Reacts {
  like: number;
  haha: number;
  wow: number;
  sad: number;
}

export interface PublicPost extends Post {
  isVenting: boolean;
  reacts: Reacts;
}
