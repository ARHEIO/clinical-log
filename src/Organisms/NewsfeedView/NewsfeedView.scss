.newsfeed-container {
  h2 {
    margin: 16px;

    @media (min-width: 960px) {
      margin: 24px;
      margin-left: 0;
    }
  }
}
