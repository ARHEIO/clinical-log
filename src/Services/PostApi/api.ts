/* eslint-disable max-len */
/* eslint-disable object-curly-newline */
import { PublicPost, Post } from './models';

const publicPosts: PublicPost[] = [
  {
    name: {
      firstName: 'Enzo',
      lastName: 'Brooks',
    },
    time: new Date().toString(),
    content: 'Up to one-third of patients experience markers of distress during palliative ventilator withdrawal, study finds',
    isVenting: false,
    reacts: { wow: 12, haha: 0, like: 5, sad: 0 },
  },
  {
    name: {
      firstName: 'Chris',
      lastName: 'Redfield',
    },
    time: '2020-06-15T03:40:44.816Z',
    content: 'Scientists have discovered a population of neurons that control hibernation-like behavior, or torpor, in mice',
    isVenting: false,
    reacts: { wow: 0, haha: 0, like: 12, sad: 0 },
  },
  {
    name: {
      firstName: 'Chris',
      lastName: 'Redfield',
    },
    time: '2020-06-14T03:40:44.816Z',
    content: 'Scientists have discovered a population of neurons that control hibernation-like behavior, or torpor, in mice',
    isVenting: false,
    reacts: { wow: 0, haha: 0, like: 12, sad: 0 },
  },
  {
    name: {
      firstName: 'Enzo',
      lastName: 'Brooks',
    },
    time: '2020-06-12T01:30:35.096Z',
    content: 'Today, @Jerry is officially cancer free!',
    isVenting: true,
    reacts: { wow: 0, haha: 0, like: 12, sad: 0 },
  },
];
const clinicalLogs: Post[] = [
  {
    name: {
      firstName: 'Chris',
      lastName: 'Redfield',
    },
    time: '2020-06-12T01:30:35.096Z',
    content: 'Patient in room 78 blood pressure is 128/70',
  },
  {
    name: {
      firstName: 'Samantha',
      lastName: 'Yule',
    },
    time: '2020-06-12T01:30:35.096Z',
    content: 'Patient in room 47 is showing decreased signs of self awareness',
  },
  {
    name: {
      firstName: 'Samantha',
      lastName: 'Yule',
    },
    time: '2020-06-12T01:30:35.096Z',
    content: 'Patient in room 78 blood pressure is 134/96',
  },
  {
    name: {
      firstName: 'Enzo',
      lastName: 'Brooks',
    },
    time: '2020-06-12T01:30:35.096Z',
    content: 'Patient in room 78 blood pressure is 139/104',
  },
  {
    name: {
      firstName: 'Samantha',
      lastName: 'Yule',
    },
    time: '2020-06-12T01:30:35.096Z',
    content: 'Patient in room 78 blood pressure is 119/79',
  },
];

const successfulCall = (): boolean => (Math.round(Math.random() * 100)) < 95; // simulating a network failure

export const getPublicPosts = (): Promise<PublicPost[]> => new Promise((resolve, reject) => setTimeout(() => (successfulCall() ? resolve(publicPosts) : reject()), 1000));

export const getClinicalLog = (): Promise<Post[]> => new Promise((resolve, reject) => setTimeout(() => (successfulCall() ? resolve(clinicalLogs) : reject()), 1000));
