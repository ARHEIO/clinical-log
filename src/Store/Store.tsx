import React, { ReactElement, useState, createContext } from 'react';
import { ThemeProvider, createMuiTheme, Theme } from '@material-ui/core/styles';

import createBreakpoints from '@material-ui/core/styles/createBreakpoints';

/**
 * xs, extra-small: 0px
 * sm, small: 600px
 * md, medium: 960px
 * lg, large: 1280px
 * xl, extra-large: 1920px
 */
const breakpoints = createBreakpoints({});

const getPixelSize = (x: number): string => `${x * 8}px`;

const theme: Theme = createMuiTheme({
  palette: {
    primary: {
      main: '#546A7B',
    },
    secondary: {
      main: '#87AEB5',
    },
  },
  overrides: {
    MuiButton: {
      root: {
        fontSize: '14px',
        minWidth: '100%',
        minHeight: getPixelSize(6),
        [breakpoints.up('md')]: {
          minWidth: getPixelSize(25),
        },
      },
    },
  },
});

export const AuthContext = createContext<any>({ isAuthed: false });

const Store = (props: {children: any}): ReactElement => {
  const { children } = props;
  const [authContext, setAuthContext] = useState({ isAuthed: false });

  return (
    <>
      <AuthContext.Provider value={[authContext, setAuthContext]}>
        <ThemeProvider theme={theme}>
          {children}
        </ThemeProvider>
      </AuthContext.Provider>
    </>
  );
};

export default Store;
