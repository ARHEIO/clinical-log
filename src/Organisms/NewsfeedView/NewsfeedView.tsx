/* eslint-disable react/no-array-index-key */
import React, {
  useEffect, useState, ReactElement, useContext,
} from 'react';
import './NewsfeedView.scss';
import { Button } from '@material-ui/core';
import Spinner from '../../Atoms/Spinner/Spinner';
import Card from '../../Atoms/Card/Card';
import PublicPostCard from '../../Molecules/PublicPostCard/PublicPostCard';
import Toolbar from '../../Molecules/Toolbar/Toolbar';
import { AuthContext } from '../../Store/Store';
import { getPublicPosts } from '../../Services/PostApi/api';
import { PublicPost } from '../../Services/PostApi/models';

const NewsfeedView = (): ReactElement => {
  const [publicPosts, setPublicPosts] = useState<PublicPost[] | null>(null);
  const [error, setError] = useState<string | null>(null);
  const [authState] = useContext(AuthContext);

  const getPosts = (): void => {
    getPublicPosts()
      .then((posts: PublicPost[]) => setPublicPosts(posts))
      .catch(() => setError('Failed to load posts'));
  };

  useEffect(() => {
    getPosts();
  }, []);

  // TODO find a better way to handle errors than nested statements
  // TODO make global error panel
  return error
    ? (
      <Card>
        {error}
        {' '}
        <Button variant="outlined" color="secondary" onClick={getPosts}>Try Again</Button>
      </Card>
    )
    : (
      <div className="newsfeed-container">
        { authState.isAuthed && (
          <>
            <h2>Authed Toolbar</h2>
            <Card><Toolbar location="newsfeed" /></Card>
          </>
        )}
        <h2>Newsfeed</h2>
        { publicPosts
          ? (
            <div className="post-container">
              {publicPosts && publicPosts.map((post: PublicPost, index) => <PublicPostCard key={`post_${index}`} post={post} />)}
            </div>
          )
          : <Spinner />}
      </div>
    );
};

export default NewsfeedView;
