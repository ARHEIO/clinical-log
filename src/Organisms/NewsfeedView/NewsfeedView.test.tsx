import React, { ReactElement } from 'react';
import { render, RenderResult, wait } from '@testing-library/react';
import pretty from 'pretty';
import NewsfeedView from './NewsfeedView';
import { AuthContext } from '../../Store/Store';
import * as endpoints from '../../Services/PostApi/api';
import { PublicPost } from '../../Services/PostApi/models';

jest.mock('../../Molecules/PublicPostCard/PublicPostCard', () => (): ReactElement => (<p>PublicPostCard</p>));

const renderComponent = (isAuthed = false): RenderResult => (
  render(
    <AuthContext.Provider value={[isAuthed]}>
      <NewsfeedView />
    </AuthContext.Provider>,
  )
);
const exampleLog: PublicPost = {
  name: {
    firstName: 'Chris',
    lastName: 'Redfield',
  },
  time: new Date().toISOString(),
  content: 'Scientists have discovered a population of neurons that control hibernation-like behavior, or torpor, in mice',
  isVenting: false,
  reacts: {
    wow: 0, haha: 0, like: 12, sad: 0,
  },
};

describe('Newsfeed view view', () => {
  it('should render', async () => {
    jest.spyOn(endpoints, 'getPublicPosts').mockImplementation(jest.fn(() => Promise.resolve([exampleLog])));
    const { container } = renderComponent();
    await wait(() => {
      expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
  });

  it('should show an error if the api call fails', async () => {
    jest.spyOn(endpoints, 'getPublicPosts').mockImplementation(jest.fn(() => Promise.reject()));
    const { container } = renderComponent();
    await wait(() => {
      expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
  });
});
