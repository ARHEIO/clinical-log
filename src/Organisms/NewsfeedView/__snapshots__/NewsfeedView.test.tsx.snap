// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsfeed view view should render 1`] = `
"<div class=\\"newsfeed-container\\">
  <h2>Newsfeed</h2>
  <div class=\\"post-container\\">
    <div class=\\"card-container public-post\\">
      <div class=\\"public-post-top\\">
        <div class=\\"public-post-top__name\\">
          <div class=\\"icon-container\\">
            <p>CR</p>
          </div>
          <p>Chris Redfield</p>
        </div>
        <div class=\\"public-post-top__date\\">
          <p>2 hours ago</p>
        </div>
      </div>
      <p class=\\"public-post__content\\">Scientists have discovered a population of neurons that control hibernation-like behavior, or torpor, in mice</p>
      <div class=\\"reacts\\"><span class=\\"react-entity\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorSecondary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z\\"></path></svg><p>12</p></span><span class=\\"react-entity\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorSecondary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm3.5-9c.83 0 1.5-.67 1.5-1.5S16.33 8 15.5 8 14 8.67 14 9.5s.67 1.5 1.5 1.5zm-7 0c.83 0 1.5-.67 1.5-1.5S9.33 8 8.5 8 7 8.67 7 9.5 7.67 11 8.5 11zm3.5 6.5c2.33 0 4.31-1.46 5.11-3.5H6.89c.8 2.04 2.78 3.5 5.11 3.5z\\"></path></svg><p>0</p></span><span class=\\"react-entity\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorSecondary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M7 13c1.66 0 3-1.34 3-3S8.66 7 7 7s-3 1.34-3 3 1.34 3 3 3zm12-6h-8v7H3V5H1v15h2v-3h18v3h2v-9c0-2.21-1.79-4-4-4z\\"></path></svg><p>0</p></span><span class=\\"react-entity\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorSecondary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><circle cx=\\"15.5\\" cy=\\"9.5\\" r=\\"1.5\\"></circle><circle cx=\\"8.5\\" cy=\\"9.5\\" r=\\"1.5\\"></circle><path d=\\"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-6c-2.33 0-4.32 1.45-5.12 3.5h1.67c.69-1.19 1.97-2 3.45-2s2.75.81 3.45 2h1.67c-.8-2.05-2.79-3.5-5.12-3.5z\\"></path></svg><p>0</p></span><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div>
    </div>
  </div>
</div>"
`;

exports[`Newsfeed view view should show an error if the api call fails 1`] = `"<div class=\\"card-container\\">Failed to load posts <button class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Try Again</span><span class=\\"MuiTouchRipple-root\\"></span></button></div>"`;
